feat(opds): implement full internationalization and refactor feed gen…

zeedif opened the pull request:
This commit introduces a comprehensive internationalization (i18n) framework and significantly refactors the OPDS v1.2 implementation for improved robustness, spec compliance, and localization.

Key changes:

### Internationalization (`i18n`):
- Introduces `LocalizationService` to manage translations:
  - Loads localized strings from JSON files (e.g., `en.json`, `es.json`) stored in a new `i18n` data directory.
  - Default `en.json` and `es.json` files are bundled and copied from resources on first run if not present.
  - Supports template resolution with `$t()` cross-references, locale fallbacks (to “en” by default), and argument interpolation (`{{placeholder}}`).
- `ServerSetup` now initializes the `i18n` directory and `LocalizationService`.

### OPDS Refactor & Enhancements:
- Replaces the previous `Opds.kt` and `OpdsDataClass.kt` with a new `OpdsFeedBuilder.kt` and a set of more granular, spec-aligned XML models (e.g., `OpdsFeedXml`, `OpdsEntryXml`, `OpdsLinkXml`).
- Integrates `LocalizationService` throughout all OPDS feeds:
  - All user-facing text (feed titles, entry titles, summaries, link titles, facet labels for sorting/filtering) is now localized.
  - Adds a `lang` query parameter to all OPDS endpoints to allow clients to request a specific UI language.
  - Uses the `Accept-Language` header as a fallback for language detection.
- The OpenSearch description (`/search` endpoint) is now localized and its template URL includes the determined language.
- Centralizes OPDS constants (namespaces, link relations, media types) in `OpdsConstants.kt`.
- Adds utility classes `OpdsDateUtil.kt`, `OpdsStringUtil.kt`, and `OpdsXmlUtil.kt` for common OPDS tasks.
- `MangaDataClass` now includes `sourceLang` to provide the content language of the manga in OPDS entries (`<dc:language>`).
- Updates OpenAPI documentation for OPDS endpoints with more detail and includes the new `lang` parameter.

### Configuration:
- Adds `useBinaryFileSizes` server configuration option. File sizes in OPDS feeds now respect this setting (e.g., MiB vs MB), utilized via `OpdsStringUtil.formatFileSizeForOpds`.

This major refactor addresses the request for internationalization originally mentioned in PR #1257 (“it would be great if messages were adapted based on the user’s language settings”). It builds upon the foundational OPDS work in #1257 and subsequent enhancements in #1262, #1263, #1278, and #1392, providing a more stable and extensible OPDS implementation. Features like localized facet titles from #1392 are now fully integrated with the i18n system.

This resolves long-standing requests for better OPDS support (e.g., issue #769) by making feeds more user-friendly, accessible, and standards-compliant, also improving the robustness of features requested in #1390 (resolved by #1392) and addressing underlying data needs for issues like #1265 (related to #1277, #1278).

zeedif committed:
feat(opds): implement full internationalization and refactor feed generation

This commit introduces a comprehensive internationalization (i18n) framework
and significantly refactors the OPDS v1.2 implementation for improved
robustness, spec compliance, and localization.

Key changes:

Internationalization (`i18n`):
- Introduces `LocalizationService` to manage translations:
    - Loads localized strings from JSON files (e.g., `en.json`, `es.json`)
      stored in a new `i18n` data directory.
    - Default `en.json` and `es.json` files are bundled and copied from
      resources on first run if not present.
    - Supports template resolution with `$t()` cross-references, locale
      fallbacks (to "en" by default), and argument interpolation ({{placeholder}}).
- `ServerSetup` now initializes the `i18n` directory and `LocalizationService`.

OPDS Refactor & Enhancements:
- Replaces the previous `Opds.kt` and `OpdsDataClass.kt` with a new
  `OpdsFeedBuilder.kt` and a set of more granular, spec-aligned XML
  models (e.g., `OpdsFeedXml`, `OpdsEntryXml`, `OpdsLinkXml`).
- Integrates `LocalizationService` throughout all OPDS feeds:
    - All user-facing text (feed titles, entry titles, summaries,
      link titles, facet labels for sorting/filtering) is now localized.
    - Adds a `lang` query parameter to all OPDS endpoints to allow
      clients to request a specific UI language.
    - Uses the `Accept-Language` header as a fallback for language detection.
- The OpenSearch description (`/search` endpoint) is now localized and
  its template URL includes the determined language.
- Centralizes OPDS constants (namespaces, link relations, media types)
  in `OpdsConstants.kt`.
- Adds utility classes `OpdsDateUtil.kt`, `OpdsStringUtil.kt`, and
  `OpdsXmlUtil.kt` for common OPDS tasks.
- `MangaDataClass` now includes `sourceLang` to provide the content
  language of the manga in OPDS entries (`<dc:language>`).
- Updates OpenAPI documentation for OPDS endpoints with more detail
  and includes the new `lang` parameter.

Configuration:
- Adds `useBinaryFileSizes` server configuration option. File sizes in
  OPDS feeds now respect this setting (e.g., MiB vs MB), utilized via
  `OpdsStringUtil.formatFileSizeForOpds`.

This major refactor addresses the request for internationalization
originally mentioned in PR #1257 ("it would be great if messages were
adapted based on the user's language settings"). It builds upon the
foundational OPDS work in #1257 and subsequent enhancements in #1262,
#1263, #1278, and #1392, providing a more stable and extensible
OPDS implementation. Features like localized facet titles from #1392
are now fully integrated with the i18n system.

This resolves long-standing requests for better OPDS support (e.g., issue #769)
by making feeds more user-friendly, accessible, and standards-compliant,
also improving the robustness of features requested in #1390 (resolved by #1392)
and addressing underlying data needs for issues like #1265 (related to #1277, #1278).

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/i18n/LocalizationService.kt:
Hmm, I'm not sure about implementing everything on our own. I would personally prefer to use a known Library like Moko Resources for localization, that would also give us some compile-time safety

zeedif replied:
@Syer10, I understand the preference for using a well-established library like Moko Resources for i18n, especially regarding its compile-time safety and potential for broader multiplatform support.

My goal with this PR was to deliver an enhancement to the OPDS support, and since I'm still familiarizing myself with the Kotlin ecosystem and its libraries, I opted for a direct implementation for the i18n part to get the functionality operational and provide the localized strings. The current `LocalizationService` already offers a foundational structure using JSON files for English and Spanish, and supports key features like template resolution (`$t()`), locale fallbacks, and argument interpolation.

Beyond the internationalization aspect, this PR also introduces other significant improvements to the OPDS functionality:

*   A complete refactor of the OPDS feed generation logic.
*   Full implementation of OPDS facets via additional link relations for chapter sorting (e.g., by date, by chapter number, ascending/descending) and filtering (e.g., read/unread), as discussed in issues like #1390.
*   Addition of `sourceLang` to `MangaDataClass` to correctly populate the `<dc:language>` tag in OPDS entries, indicating the content language of the manga.

I have no issue with the i18n part being migrated to Moko Resources if that's what you prefer. The existing JSON translation files and the logic for determining user language preference can surely help you in working on that.

Syer10 replied:
Got it, if you need me to I can work on getting a Moko implementation within the next few days added to this PR.

zeedif replied:
@Syer10 I've been looking into Moko Resources and tried to understand how to integrate it here, but I couldn't quite wrap my head around the setup, so I'd prefer if you could try to implement it. I'm sure you'll have a much clearer path to get it working smoothly, and thanks in advance!

zeedif committed:
fix(opds): revert MIME type to application/xml for browser compatibility

zeedif committed:
fix(opds): use chapter index for metadata feed and correct link relation

- Change `getChapterMetadataFeed` to use `chapterIndexFromPath` (sourceOrder)
  instead of `chapterIdFromPath` for fetching chapter data, ensuring
  consistency with how chapters are identified in manga feeds.
- Add error handling for cases where manga or chapter by index is not found.
- Correct OPDS link relation for chapter detail/fetch link in non-metadata
  chapter entries from `alternate` to `subsection` as per OPDS spec
  for navigation to more specific content or views.

Syer10 committed:
Use Moko-Resources

Syer10 committed:
Format

Syer10 committed:
Forgot the Languages.json

zeedif committed:
refactor(opds)!: restructure OPDS feeds and introduce data repositories

This commit significantly refactors the OPDS v1.2 implementation by introducing dedicated repository classes for data fetching and by restructuring the feed generation logic for clarity and maintainability. The `chapterId` path parameter for chapter metadata feeds has been changed to `chapterIndex` (sourceOrder) to align with how chapters are identified in manga feeds.

BREAKING CHANGE: The OPDS endpoint for chapter metadata has changed from `/api/opds/v1.2/manga/{mangaId}/chapter/{chapterId}/fetch` to `/api/opds/v1.2/manga/{mangaId}/chapter/{chapterIndex}/fetch`. Clients will need to update to use the chapter's source order (index) instead of its database ID.

Key changes:
- Introduced `MangaRepository`, `ChapterRepository`, and `NavigationRepository` to encapsulate database queries and data transformation logic for OPDS feeds.
- Moved data fetching logic from `OpdsFeedBuilder` to these new repositories.
- `OpdsFeedBuilder` now primarily focuses on constructing the XML feed structure using DTOs provided by the repositories.
- Renamed `OpdsMangaAcqEntry.thumbnailUrl` to `rawThumbnailUrl` for clarity.
- Added various DTOs (e.g., `OpdsRootNavEntry`, `OpdsMangaDetails`, `OpdsChapterListAcqEntry`) to define clear data contracts between repositories and the feed builder.
- Simplified `OpdsV1Controller` by reorganizing feed endpoints into logical groups (Main Navigation, Filtered Acquisition, Item-Specific).
- Updated `OpdsAPI` to reflect the path parameter change for chapter metadata (`chapterIndex` instead of `chapterId`).
- Added `slugify()` utility to `OpdsStringUtil` for creating URL-friendly genre IDs.
- Standardized localization keys for root feed entry descriptions to use `*.entryContent` instead of `*.description`.
- Added `server.generated.BuildConfig` (likely from build process).

zeedif committed:
style(opds): apply ktlint fixes

Syer10 commented on server/bin/main/suwayomi/tachidesk/i18n/LocalizationHelper.kt:
I think you accidentally moved server/build/bin to server/bin and commited it

zeedif committed:
Delete server/bin

zeedif committed:
Delete server/bin

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt:
```diff
@@ -222,6 +226,90 @@ fun applicationSetup() {
     // fixes #119 , ref: https://github.com/Suwayomi/Suwayomi-Server/issues/119#issuecomment-894681292 , source Id calculation depends on String.lowercase()
     Locale.setDefault(Locale.ENGLISH)
 
+// Make sure the localization directory exists
```
I don't see the point of this giant block of code, what is it for?

zeedif replied:
@Syer10 Whoops, missed that during the merge with the Moko changes. That initialization is no longer needed. Removing it now.

zeedif committed:
refactor(i18n): remove custom LocalizationService initialization

zeedif committed:
refactor(i18n): remove unused imports from ServerSetup

zeedif committed:
refactor(model): remove sourceLang from MangaDataClass

shirishsaxena commented on server/src/main/kotlin/suwayomi/tachidesk/opds/impl/OpdsFeedBuilder.kt:
```diff
@@ -0,0 +1,1016 @@
+        val builder =
+            FeedBuilderInternal(
+                baseUrl = baseUrl,
+                idPath = "manga/$mangaId/chapters",
```
idPath seems to point to an invalid endpoint (pagination doesn't work anymore)
```suggestion
                idPath = "manga/$mangaId",
```

zeedif replied:
Those path changes were part of the OPDS API restructuring I forgot to push earlier.

shirishsaxena commented on server/src/main/resources/server-reference.conf:
```diff
@@ -72,6 +72,7 @@ server.flareSolverrSessionTtl = 15 # time in minutes
 server.flareSolverrAsResponseFallback = false
 
 # OPDS
+server.useBinaryFileSizes = false # if the file sizes should be displayed in binary (KiB, MiB, GiB) or decimal (KB, MB, GB)
```
Just a nitpick but if this is just for OPDS, can we refactor it to `server.opdsUseBinaryFileSizes`?

zeedif committed:
refactor(opds): rename OPDS binary file size config property

- Rename `useBinaryFileSizes` to `opdsUseBinaryFileSizes` in code and config
- Update related condition check in formatFileSizeForOpds

BREAKING CHANGE: Existing server configurations using `server.useBinaryFileSizes` need to migrate to `server.opdsUseBinaryFileSizes`

zeedif committed:
refactor(opds): improve OPDS endpoint structure and documentation

- Restructure endpoint paths for better resource hierarchy
- Add descriptive comments for each feed type and purpose
- Rename `/fetch` endpoint to `/metadata` for clarity
- Standardize feed naming conventions in route definitions

BREAKING CHANGE: Existing OPDS client integrations using old endpoint paths (`/manga/{mangaId}` and `/chapter/{chapterIndex}/fetch`) require updates to new paths (`/manga/{mangaId}/chapters` and `/chapter/{chapterIndex}/metadata`)

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/i18n/LocalizationHelper.kt:
```diff
@@ -0,0 +1,48 @@
+private val logger = KotlinLogging.logger {}
+
+object LocalizationHelper {
+    // Supported language codes (lowercase)
```
Accidentally left this

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/opds/impl/OpdsFeedBuilder.kt:
```diff
@@ -0,0 +1,1016 @@
+        val feedTitle =
+            MR.strings.opds_feeds_source_specific_title.localized(
+                locale,
+                sourceNavEntry?.name ?: MR.strings.opds_feeds_source_specific_title.localized(locale, sourceId.toString()),
```
If you are going to pass a numer, set the string xml to take a number instead of a string. Make sure to update both the translated files

Here is the relavent documentation:
https://developer.android.com/reference/kotlin/java/util/Formatter.html#conversions

Though now that I look again, it looks like a logic error instead of a improperly passed value. It looks like its getting localized 2 times

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/opds/impl/OpdsFeedBuilder.kt:
```diff
@@ -0,0 +1,1016 @@
+                ?: return buildNotFoundFeed(
+                    baseUrl,
+                    "manga/$mangaId",
+                    MR.strings.opds_error_manga_not_found.localized(locale, mangaId.toString()),
```
Same here

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/opds/impl/OpdsFeedBuilder.kt:
```diff
@@ -0,0 +1,1016 @@
+                ?: return buildNotFoundFeed(
+                    baseUrl,
+                    "manga/$mangaId/chapter/$chapterSourceOrder/metadata",
+                    MR.strings.opds_error_manga_not_found.localized(locale, mangaId.toString()),
```
Here

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/opds/impl/OpdsFeedBuilder.kt:
```diff
@@ -0,0 +1,1016 @@
+                ?: return buildNotFoundFeed(
+                    baseUrl,
+                    "manga/$mangaId/chapter/$chapterSourceOrder/metadata",
+                    MR.strings.opds_error_chapter_not_found.localized(locale, chapterSourceOrder.toString()),
```
Here

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/opds/repository/ChapterRepository.kt:
```diff
@@ -0,0 +1,131 @@
+                    .select(
+                        ChapterTable.columns + MangaTable.title + MangaTable.author + MangaTable.thumbnail_url + MangaTable.id +
+                            SourceTable.lang,
+                    ) // Reemplaza slice().selectAll()
```
English docs

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/opds/repository/ChapterRepository.kt:
```diff
@@ -0,0 +1,131 @@
+            val items =
+                query
+                    .orderBy(ChapterTable.fetchedAt to SortOrder.DESC, ChapterTable.sourceOrder to SortOrder.DESC)
+                    .limit(opdsItemsPerPageBounded) // Moderno: .limit(n).offset(m)
```
And here

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/opds/repository/ChapterRepository.kt:
```diff
@@ -0,0 +1,131 @@
+                    .offset(((pageNum - 1) * opdsItemsPerPageBounded).toLong())
+                    .map {
+                        OpdsLibraryUpdateAcqEntry(
+                            chapter = it.toOpdsChapterListAcqEntry(), // Esto funcionará si las columnas de ChapterTable no colisionan
```
And here

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/opds/repository/NavigationRepository.kt:
```diff
@@ -0,0 +1,210 @@
+    private val opdsItemsPerPageBounded: Int
+        get() = serverConfig.opdsItemsPerPage.value.coerceIn(10, 5000)
+
+    // Mapeo de IDs de sección a sus StringResources para título y descripción
```
English docs

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/opds/repository/NavigationRepository.kt:
```diff
@@ -0,0 +1,210 @@
+        }
+
+    fun getStatuses(locale: Locale): List<OpdsStatusNavEntry> {
+        // Mapeo de MangaStatus a sus StringResources
```
And here

zeedif committed:
fix(opds): Apply review suggestions for localization and comments

schroda commented on server/i18n/src/commonMain/moko-resources/values/es/strings.xml:
should we create a weblate component for handling the non english translations?

Syer10 replied:
Yes, we can that once this is merged. Weblate can start with both the english and spanish, its just after that we can't do manual translations.

zeedif committed:
Fix

zeedif committed:
fix(opds): Update chapter links to include 'chapters' and 'metadata' in URLs
